# 01. Importing libraries

In [30]:
# import libraries
import pandas as pd
import numpy as np
import os

In [31]:
path = r'/Users/kirstinhelgeson/Desktop/Data Analytics Course/07-2023 Achievement 6 Project'

In [32]:
# import dataframe
df_conditions = pd.read_csv(os.path.join(path,'02 Data','Original Data','Conditions_Contributing_to_COVID-19_Deaths__by_State_and_Age__Provisional_2020-2023.csv'), index_col = False)

# 02. Explore the Data

In [33]:
# check first 5 entries of original dataframe
df_conditions.head()

Data As Of  Start Date    End Date     Group  Year  Month          State  \
0  06/25/2023  01/01/2020  06/24/2023  By Total   NaN    NaN  United States   
1  06/25/2023  01/01/2020  06/24/2023  By Total   NaN    NaN  United States   
2  06/25/2023  01/01/2020  06/24/2023  By Total   NaN    NaN  United States   
3  06/25/2023  01/01/2020  06/24/2023  By Total   NaN    NaN  United States   
4  06/25/2023  01/01/2020  06/24/2023  By Total   NaN    NaN  United States   

        Condition Group                Condition ICD10_codes Age Group  \
0  Respiratory diseases  Influenza and pneumonia     J09-J18      0-24   
1  Respiratory diseases  Influenza and pneumonia     J09-J18     25-34   
2  Respiratory diseases  Influenza and pneumonia     J09-J18     35-44   
3  Respiratory diseases  Influenza and pneumonia     J09-J18     45-54   
4  Respiratory diseases  Influenza and pneumonia     J09-J18     55-64   

   COVID-19 Deaths  Number of Mentions Flag  
0           1554.0              1630.0  NaN  
1           5775.0              5998.0  NaN  
2          15026.0             15643.0  NaN  
3          37335.0             38794.0  NaN  
4          82382.0             85404.0  NaN

In [34]:
# check last 5 entries of original dataframe
df_conditions.tail()

Data As Of  Start Date    End Date     Group    Year  Month  \
583735  06/25/2023  02/01/2023  02/28/2023  By Month  2023.0    2.0   
583736  06/25/2023  03/01/2023  03/31/2023  By Month  2023.0    3.0   
583737  06/25/2023  04/01/2023  04/30/2023  By Month  2023.0    4.0   
583738  06/25/2023  05/01/2023  05/31/2023  By Month  2023.0    5.0   
583739  06/25/2023  06/01/2023  06/24/2023  By Month  2023.0    6.0   

              State Condition Group Condition ICD10_codes Age Group  \
583735  Puerto Rico        COVID-19  COVID-19        U071  All Ages   
583736  Puerto Rico        COVID-19  COVID-19        U071  All Ages   
583737  Puerto Rico        COVID-19  COVID-19        U071  All Ages   
583738  Puerto Rico        COVID-19  COVID-19        U071  All Ages   
583739  Puerto Rico        COVID-19  COVID-19        U071  All Ages   

        COVID-19 Deaths  Number of Mentions Flag  
583735             99.0                99.0  NaN  
583736             50.0                50.0  NaN  
583737             44.0                44.0  NaN  
583738             65.0                65.0  NaN  
583739             71.0                71.0  NaN

In [35]:
# check shape of original dataframe
df_conditions.shape

(583740, 14)

In [36]:
# check basic info on dataframe columns
df_conditions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 583740 entries, 0 to 583739
Data columns (total 14 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   Data As Of          583740 non-null  object 
 1   Start Date          583740 non-null  object 
 2   End Date            583740 non-null  object 
 3   Group               583740 non-null  object 
 4   Year                571320 non-null  float64
 5   Month               521640 non-null  float64
 6   State               583740 non-null  object 
 7   Condition Group     583740 non-null  object 
 8   Condition           583740 non-null  object 
 9   ICD10_codes         583740 non-null  object 
 10  Age Group           583740 non-null  object 
 11  COVID-19 Deaths     411171 non-null  float64
 12  Number of Mentions  416761 non-null  float64
 13  Flag                172569 non-null  object 
dtypes: float64(4), object(10)
memory usage: 62.4+ MB


### Data types are all consistent, but there are null values in a few columns that need to be further examined.

In [37]:
# check list of column names in original dataframe
df_conditions.columns

Index(['Data As Of', 'Start Date', 'End Date', 'Group', 'Year', 'Month',
       'State', 'Condition Group', 'Condition', 'ICD10_codes', 'Age Group',
       'COVID-19 Deaths', 'Number of Mentions', 'Flag'],
      dtype='object')

# 03. Describe the Data

In [38]:
df_conditions.describe()

Year          Month  COVID-19 Deaths  Number of Mentions
count  571320.000000  521640.000000     4.111710e+05        4.167610e+05
mean     2021.304348       6.071429     1.266096e+02        1.362334e+02
std         1.039850       3.425349     3.052289e+03        3.279466e+03
min      2020.000000       1.000000     0.000000e+00        0.000000e+00
25%      2020.000000       3.000000     0.000000e+00        0.000000e+00
50%      2021.000000       6.000000     0.000000e+00        0.000000e+00
75%      2022.000000       9.000000     1.900000e+01        2.100000e+01
max      2023.000000      12.000000     1.135624e+06        1.135624e+06

### None of these values seem out of the ordinary, however it should be noted that this dataframe contains rows of data for the entire United States during all of 2020-2023 (By Total), rows of data per year (By Year) and rows of data for each month of each year (By Month) so the statistical values such as the mean may not refer to the actual average death count from 2020 to 2023. May consider separating these categories into separate dataframes to allow for better statistical analysis, especially by month.

# 04. Remove Unnecessary Columns

In [39]:
# remove 'Data As Of' column because we can just take note that this data was all updated 7/12/23.
# remove 'Start Date' column because it doesn't give us any additional information that the 'Month' column does not.
# remove 'End Date' column because it doesn't give us any additional information that the 'Month' column does not.
# remove 'ICD10_codes' column because it is redundant and not necessary for our analysis
# remove 'Number of Mentions' column because it is not relevant for our analysis

df_conditions = df_conditions.drop(columns=['Data As Of','Start Date','End Date', 'ICD10_codes', 'Number of Mentions'])

In [40]:
# check new output of dataframe after removing 3 columns
df_conditions.head()

Group  Year  Month          State       Condition Group  \
0  By Total   NaN    NaN  United States  Respiratory diseases   
1  By Total   NaN    NaN  United States  Respiratory diseases   
2  By Total   NaN    NaN  United States  Respiratory diseases   
3  By Total   NaN    NaN  United States  Respiratory diseases   
4  By Total   NaN    NaN  United States  Respiratory diseases   

                 Condition Age Group  COVID-19 Deaths Flag  
0  Influenza and pneumonia      0-24           1554.0  NaN  
1  Influenza and pneumonia     25-34           5775.0  NaN  
2  Influenza and pneumonia     35-44          15026.0  NaN  
3  Influenza and pneumonia     45-54          37335.0  NaN  
4  Influenza and pneumonia     55-64          82382.0  NaN

### Remember that the 'By Total' group sums all data from 1/1/20 to 6/24/23.

# 05. Separate dataframe by 'Group' column classification

### I want a dataframe for each classification under the 'Group' column, 'By Total', 'By Year', 'By Month'. This way I can isolate the data specifically for each month from 2020-2023. It's not that useful to just see total deaths for all 3 years (the 'By Total' data) other than to show how many people were lost during the COVID-19 pandemic from 2020-2023. We would not be able to show any trends by looking at this data in conjunction with the data observed by month.

In [41]:
# create a new dataframe with only rows that contain 'By Month' classification
df_conditions_month = df_conditions[df_conditions['Group'] == 'By Month']

In [42]:
# check output of new dataframe
df_conditions_month.head()

Group    Year  Month          State       Condition Group  \
62100  By Month  2020.0    1.0  United States  Respiratory diseases   
62101  By Month  2020.0    2.0  United States  Respiratory diseases   
62102  By Month  2020.0    3.0  United States  Respiratory diseases   
62103  By Month  2020.0    4.0  United States  Respiratory diseases   
62104  By Month  2020.0    5.0  United States  Respiratory diseases   

                     Condition Age Group  COVID-19 Deaths Flag  
62100  Influenza and pneumonia      0-24              0.0  NaN  
62101  Influenza and pneumonia      0-24              0.0  NaN  
62102  Influenza and pneumonia      0-24              9.0  NaN  
62103  Influenza and pneumonia      0-24             27.0  NaN  
62104  Influenza and pneumonia      0-24             19.0  NaN

In [43]:
# check shape of new dataframe
df_conditions_month.shape

(521640, 9)

In [44]:
# check descriptive data of new dataframe
df_conditions_month.describe()

Year          Month  COVID-19 Deaths
count  521640.000000  521640.000000    362188.000000
mean     2021.285714       6.071429        46.996242
std         1.030158       3.425349       599.569607
min      2020.000000       1.000000         0.000000
25%      2020.000000       3.000000         0.000000
50%      2021.000000       6.000000         0.000000
75%      2022.000000       9.000000        13.000000
max      2023.000000      12.000000    105566.000000

In [45]:
# check info of new dataframe
df_conditions_month.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 521640 entries, 62100 to 583739
Data columns (total 9 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   Group            521640 non-null  object 
 1   Year             521640 non-null  float64
 2   Month            521640 non-null  float64
 3   State            521640 non-null  object 
 4   Condition Group  521640 non-null  object 
 5   Condition        521640 non-null  object 
 6   Age Group        521640 non-null  object 
 7   COVID-19 Deaths  362188 non-null  float64
 8   Flag             159452 non-null  object 
dtypes: float64(3), object(6)
memory usage: 39.8+ MB


# 06. Remove 'Group' column

In [46]:
# remove 'group' column now since it no longer provides any meaningful information
df_conditions_month_2 = df_conditions_month.drop('Group', axis=1)

In [47]:
# check output of new dataframe
df_conditions_month_2

Year  Month          State       Condition Group  \
62100   2020.0    1.0  United States  Respiratory diseases   
62101   2020.0    2.0  United States  Respiratory diseases   
62102   2020.0    3.0  United States  Respiratory diseases   
62103   2020.0    4.0  United States  Respiratory diseases   
62104   2020.0    5.0  United States  Respiratory diseases   
...        ...    ...            ...                   ...   
583735  2023.0    2.0    Puerto Rico              COVID-19   
583736  2023.0    3.0    Puerto Rico              COVID-19   
583737  2023.0    4.0    Puerto Rico              COVID-19   
583738  2023.0    5.0    Puerto Rico              COVID-19   
583739  2023.0    6.0    Puerto Rico              COVID-19   

                      Condition Age Group  COVID-19 Deaths Flag  
62100   Influenza and pneumonia      0-24              0.0  NaN  
62101   Influenza and pneumonia      0-24              0.0  NaN  
62102   Influenza and pneumonia      0-24              9.0  NaN  
62103   Influenza and pneumonia      0-24             27.0  NaN  
62104   Influenza and pneumonia      0-24             19.0  NaN  
...                         ...       ...              ...  ...  
583735                 COVID-19  All Ages             99.0  NaN  
583736                 COVID-19  All Ages             50.0  NaN  
583737                 COVID-19  All Ages             44.0  NaN  
583738                 COVID-19  All Ages             65.0  NaN  
583739                 COVID-19  All Ages             71.0  NaN  

[521640 rows x 8 columns]

In [48]:
# check shape of new dataframe
df_conditions_month_2.shape

(521640, 8)

In [49]:
# check descriptive data of new dataframe
df_conditions_month_2.describe()

Year          Month  COVID-19 Deaths
count  521640.000000  521640.000000    362188.000000
mean     2021.285714       6.071429        46.996242
std         1.030158       3.425349       599.569607
min      2020.000000       1.000000         0.000000
25%      2020.000000       3.000000         0.000000
50%      2021.000000       6.000000         0.000000
75%      2022.000000       9.000000        13.000000
max      2023.000000      12.000000    105566.000000

In [50]:
# check info of new dataframe
df_conditions_month_2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 521640 entries, 62100 to 583739
Data columns (total 8 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   Year             521640 non-null  float64
 1   Month            521640 non-null  float64
 2   State            521640 non-null  object 
 3   Condition Group  521640 non-null  object 
 4   Condition        521640 non-null  object 
 5   Age Group        521640 non-null  object 
 6   COVID-19 Deaths  362188 non-null  float64
 7   Flag             159452 non-null  object 
dtypes: float64(3), object(5)
memory usage: 35.8+ MB


# 07. Remove rows with 'United States' classification

### I want a dataframe that does not have extra values or redundant information. If I have all of the information on COVID deaths by month and by state, then I can add these values together to get US totals if I need to. Since we are going to attempt a geographical analysis though I need the information to be about States rather than the entire country.

In [51]:
# create a new dataframe with only rows that don't contain 'United States' classification
df_conditions_month_3 = df_conditions_month_2[df_conditions_month_2['State'] != 'United States']

In [52]:
# check output of new dataframe
df_conditions_month_3

Year  Month        State       Condition Group  \
71760   2020.0    1.0      Alabama  Respiratory diseases   
71761   2020.0    2.0      Alabama  Respiratory diseases   
71762   2020.0    3.0      Alabama  Respiratory diseases   
71763   2020.0    4.0      Alabama  Respiratory diseases   
71764   2020.0    5.0      Alabama  Respiratory diseases   
...        ...    ...          ...                   ...   
583735  2023.0    2.0  Puerto Rico              COVID-19   
583736  2023.0    3.0  Puerto Rico              COVID-19   
583737  2023.0    4.0  Puerto Rico              COVID-19   
583738  2023.0    5.0  Puerto Rico              COVID-19   
583739  2023.0    6.0  Puerto Rico              COVID-19   

                      Condition Age Group  COVID-19 Deaths Flag  
71760   Influenza and pneumonia      0-24              0.0  NaN  
71761   Influenza and pneumonia      0-24              0.0  NaN  
71762   Influenza and pneumonia      0-24              0.0  NaN  
71763   Influenza and pneumonia      0-24              0.0  NaN  
71764   Influenza and pneumonia      0-24              0.0  NaN  
...                         ...       ...              ...  ...  
583735                 COVID-19  All Ages             99.0  NaN  
583736                 COVID-19  All Ages             50.0  NaN  
583737                 COVID-19  All Ages             44.0  NaN  
583738                 COVID-19  All Ages             65.0  NaN  
583739                 COVID-19  All Ages             71.0  NaN  

[511980 rows x 8 columns]

In [53]:
# check shape of new dataframe
df_conditions_month_3.shape

(511980, 8)

In [54]:
# check descriptive data of new dataframe
df_conditions_month_3.describe()

Year          Month  COVID-19 Deaths
count  511980.000000  511980.000000     352528.00000
mean     2021.285714       6.071429         23.47255
std         1.030159       3.425349        140.64035
min      2020.000000       1.000000          0.00000
25%      2020.000000       3.000000          0.00000
50%      2021.000000       6.000000          0.00000
75%      2022.000000       9.000000         12.00000
max      2023.000000      12.000000      20066.00000

In [55]:
# check info of new dataframe
df_conditions_month_3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 511980 entries, 71760 to 583739
Data columns (total 8 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   Year             511980 non-null  float64
 1   Month            511980 non-null  float64
 2   State            511980 non-null  object 
 3   Condition Group  511980 non-null  object 
 4   Condition        511980 non-null  object 
 5   Age Group        511980 non-null  object 
 6   COVID-19 Deaths  352528 non-null  float64
 7   Flag             159452 non-null  object 
dtypes: float64(3), object(5)
memory usage: 35.2+ MB


# 08. Missing Values

In [56]:
# check null value count for each column
df_conditions_month_3.isnull().sum()

Year                    0
Month                   0
State                   0
Condition Group         0
Condition               0
Age Group               0
COVID-19 Deaths    159452
Flag               352528
dtype: int64

### The 159452 null values for 'COVID-19 Deaths' are due to the following reason noted in the 'Flag' column: "One or more data cells have counts between 1-9 and have been suppressed in accordance with NCHS confidentiality standards." The 'Flag' column has null values where there were no necessary comments such as this.

### I think the missing values in the COVID-19 Death column can be replaced with a random integer 1-9 because we know that at least one death occurred and can't be reported because of confidentiality and privacy issues, and based on how they update this data every month the newer counts in particular are likely an underestimate of the total deaths as they catch up with incoming death certificate data, so I do not believe that the data will be skewed irreparably by doing this. It is worth noting that null values make up 31% of values in the 'COVID-19 Deaths' column.

# 09. Duplicate values

In [57]:
# create subset containing any duplicate values in the dataframe
df_dups = df_conditions_month_3[df_conditions_month_3.duplicated()]

In [58]:
# print df_dups to observe duplicates
df_dups

Empty DataFrame
Columns: [Year, Month, State, Condition Group, Condition, Age Group, COVID-19 Deaths, Flag]
Index: []

### No duplicates found.

# 10. Replace Null Values

In [59]:
# create list of columns to replace null values
columns_to_fill = ['COVID-19 Deaths']

In [60]:
# create new dataframe that has all null values replaced by random integer from 1 to 9 inclusive
df_conditions_month_replaced = df_conditions_month_3[columns_to_fill].fillna(np.random.randint(1,10))

In [61]:
# add the other columns that were not changed from df_conditions_month_3 dataframe
df_conditions_month_replaced = pd.concat([df_conditions_month_3.drop(columns=columns_to_fill), df_conditions_month_replaced], axis=1)

In [62]:
# check null values for dataframe that replaces all null values
df_conditions_month_replaced.isnull().sum()

Year                    0
Month                   0
State                   0
Condition Group         0
Condition               0
Age Group               0
Flag               352528
COVID-19 Deaths         0
dtype: int64

In [63]:
# describe new dataframe
df_conditions_month_replaced.describe()

Year          Month  COVID-19 Deaths
count  511980.000000  511980.000000    511980.000000
mean     2021.285714       6.071429        17.719425
std         1.030159       3.425349       117.015570
min      2020.000000       1.000000         0.000000
25%      2020.000000       3.000000         0.000000
50%      2021.000000       6.000000         5.000000
75%      2022.000000       9.000000         5.000000
max      2023.000000      12.000000     20066.000000

In [64]:
# check output of wrangled and cleaned dataframe
df_conditions_month_replaced

Year  Month        State       Condition Group  \
71760   2020.0    1.0      Alabama  Respiratory diseases   
71761   2020.0    2.0      Alabama  Respiratory diseases   
71762   2020.0    3.0      Alabama  Respiratory diseases   
71763   2020.0    4.0      Alabama  Respiratory diseases   
71764   2020.0    5.0      Alabama  Respiratory diseases   
...        ...    ...          ...                   ...   
583735  2023.0    2.0  Puerto Rico              COVID-19   
583736  2023.0    3.0  Puerto Rico              COVID-19   
583737  2023.0    4.0  Puerto Rico              COVID-19   
583738  2023.0    5.0  Puerto Rico              COVID-19   
583739  2023.0    6.0  Puerto Rico              COVID-19   

                      Condition Age Group Flag  COVID-19 Deaths  
71760   Influenza and pneumonia      0-24  NaN              0.0  
71761   Influenza and pneumonia      0-24  NaN              0.0  
71762   Influenza and pneumonia      0-24  NaN              0.0  
71763   Influenza and pneumonia      0-24  NaN              0.0  
71764   Influenza and pneumonia      0-24  NaN              0.0  
...                         ...       ...  ...              ...  
583735                 COVID-19  All Ages  NaN             99.0  
583736                 COVID-19  All Ages  NaN             50.0  
583737                 COVID-19  All Ages  NaN             44.0  
583738                 COVID-19  All Ages  NaN             65.0  
583739                 COVID-19  All Ages  NaN             71.0  

[511980 rows x 8 columns]

# 11. Check COVID-19 Outliers

In [65]:
# calculating quartiles
Q1 = df_conditions_month_replaced['COVID-19 Deaths'].quantile(0.25)
Q2 = df_conditions_month_replaced['COVID-19 Deaths'].quantile(0.50)
Q3 = df_conditions_month_replaced['COVID-19 Deaths'].quantile(0.75)

In [66]:
# show calculated quartiles
print("Q1:", Q1)
print("Q2:", Q2)
print("Q3:", Q3)

Q1: 0.0
Q2: 5.0
Q3: 5.0


In [67]:
# calculate standard deviation
std_dev = df_conditions_month_replaced['COVID-19 Deaths'].std()

In [68]:
# show standard deviation
std_dev

117.0155698786212

In [69]:
# calculate mean
mean = df_conditions_month_replaced['COVID-19 Deaths'].mean()

In [70]:
# show mean
mean

17.719424586897926

In [71]:
# calculate two standard deviations above mean (upper bound)
upper_bound = mean + (2 * std_dev)

In [72]:
# show upper bound
upper_bound

251.75056434414034

In [73]:
# calculate two standard deviations below mean (lower bound)
lower_bound = mean - (2 * std_dev)

In [74]:
# show lower bound
lower_bound

-216.31171517034448

### The lower bound here is negative which is not sensical when talking about population deaths, so the lower bound is effectively 0.

In [75]:
# define outliers
COVID_outliers = df_conditions_month_replaced[(df_conditions_month_replaced['COVID-19 Deaths'] < lower_bound) | (df_conditions_month_replaced['COVID-19 Deaths'] > upper_bound)]

In [76]:
# show resulting dataframe containing only COVID-19 Death outliers
COVID_outliers

Year  Month        State       Condition Group  \
72024   2021.0    1.0      Alabama  Respiratory diseases   
72144   2020.0    7.0      Alabama  Respiratory diseases   
72145   2020.0    8.0      Alabama  Respiratory diseases   
72149   2020.0   12.0      Alabama  Respiratory diseases   
72150   2021.0    1.0      Alabama  Respiratory diseases   
...        ...    ...          ...                   ...   
583708  2020.0   11.0  Puerto Rico              COVID-19   
583709  2020.0   12.0  Puerto Rico              COVID-19   
583717  2021.0    8.0  Puerto Rico              COVID-19   
583718  2021.0    9.0  Puerto Rico              COVID-19   
583722  2022.0    1.0  Puerto Rico              COVID-19   

                      Condition Age Group Flag  COVID-19 Deaths  
72024   Influenza and pneumonia     75-84  NaN            271.0  
72144   Influenza and pneumonia  All Ages  NaN            264.0  
72145   Influenza and pneumonia  All Ages  NaN            318.0  
72149   Influenza and pneumonia  All Ages  NaN            546.0  
72150   Influenza and pneumonia  All Ages  NaN            838.0  
...                         ...       ...  ...              ...  
583708                 COVID-19  All Ages  NaN            341.0  
583709                 COVID-19  All Ages  NaN            440.0  
583717                 COVID-19  All Ages  NaN            314.0  
583718                 COVID-19  All Ages  NaN            269.0  
583722                 COVID-19  All Ages  NaN            622.0  

[5875 rows x 8 columns]

In [77]:
COVID_outliers.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5875 entries, 72024 to 583722
Data columns (total 8 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Year             5875 non-null   float64
 1   Month            5875 non-null   float64
 2   State            5875 non-null   object 
 3   Condition Group  5875 non-null   object 
 4   Condition        5875 non-null   object 
 5   Age Group        5875 non-null   object 
 6   Flag             0 non-null      object 
 7   COVID-19 Deaths  5875 non-null   float64
dtypes: float64(3), object(5)
memory usage: 413.1+ KB


### There are 5875 entries of total 511980 that are considered outliers for COVID-19 Death Counts. We will leave these outliers alone since this data is sourced and collected reliably and does not indicate any error in the data.

# 12. Export wrangled dataframe

In [78]:
df_conditions_month_replaced.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'Contributing_Conditions_wrangled.csv'))